server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotTaskRunner.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.repositories.blobstore;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRunnable;
+import org.elasticsearch.common.CheckedBiConsumer;
+import org.elasticsearch.common.util.concurrent.PrioritizedThrottledTaskRunner;
+import org.elasticsearch.repositories.SnapshotShardContext;
+
+import java.io.IOException;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+
+import static org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot.FileInfo;
+
+/**
+ * {@link ShardSnapshotTaskRunner} performs snapshotting tasks, prioritizing {@link ShardSnapshotTask}
+ * over {@link FileSnapshotTask}. Each enqueued shard to snapshot results in one {@link ShardSnapshotTask}
+ * and zero or more {@link FileSnapshotTask}s.
+ */
+public class ShardSnapshotTaskRunner {
+    private final PrioritizedThrottledTaskRunner<SnapshotTask> taskRunner;
+    private final Consumer<SnapshotShardContext> shardSnapshotter;
+    private final CheckedBiConsumer<SnapshotShardContext, FileInfo, IOException> fileSnapshotter;
+
+    abstract static class SnapshotTask implements Comparable<SnapshotTask>, Runnable {
+        protected final SnapshotShardContext context;
+
+        SnapshotTask(SnapshotShardContext context) {
+            this.context = context;
+        }
+
+        public abstract short priority();
```

Comment:
Should we just use `int` here. `short` is kind of pointless isn't it, might even be a net negative for comparison performance?